doc/modules/ensemble.rst

```diff
@@ -102,24 +102,67 @@ controls the number of iterations of the boosting process::
-`n_classes >= 3`, it uses the multi-class log loss function, with multinomial deviance
-and categorical cross-entropy as alternative names. The appropriate loss version is
-selected based on :term:`y` passed to :term:`fit`.
+Available losses for regression are 'squared_error', 'absolute_error', which is
+less sensitive to outliers, and 'poisson', which is well suited to model counts
+and frequencies. For classification, 'log_loss' is the only option. For binary
+classification it uses the binary log loss, also known as binomial deviance or
+binary cross-entropy. For `n_classes >= 3`, it uses the multi-class log loss
+function, with multinomial deviance and categorical cross-entropy as alternative
+names. The appropriate loss version is selected based on :term:`y` passed to
+:term:`fit`.
 
 The size of the trees can be controlled through the ``max_leaf_nodes``,
 ``max_depth``, and ``min_samples_leaf`` parameters.
 
 The number of bins used to bin the data is controlled with the ``max_bins``
-parameter. Using less bins acts as a form of regularization. It is
-generally recommended to use as many bins as possible (256), which is the default.
+parameter. Using less bins acts as a form of regularization. It is generally
+recommended to use as many bins as possible (256), which is the default.
+
+The ``l2_regularization`` parameter acts as a regularizer for the loss function,
+and corresponds to :math:`\lambda` in the following regularized objective (see
+equation (2) of [XGBoost]_):
+
+.. math::
+
+      \mathcal{L}(\phi) =  \sum_i l(\hat{y}_i, y_i) + \sum_k \Omega(f_k) \\
+      \text{where} ~ \Omega(f_k) = \gamma T_k  + \frac12 \lambda ||w_k||^2
+
+Here :math:`l` is the loss function; :math:`T` is the number of leaves in the
+tree; :math:`f_k` corresponds to the k-th tree in the ensemble of trees (such
+that :math:`\hat{y}_i=\phi(x_i)=\sum_k f_k(x_i)`); and :math:`w_k` is a vector
+of lenght :math:`T` containing the leaf weights.
+
+Notice that :math:`\gamma` penalizes the number of leaves (which makes it a
+smooth version of `max_leaf_nodes` and is not implemented in scikit-learn),
+whereas :math:`\lambda` penalizes the magnitude of the individual tree
+predictions.
+
+**Note**: In practice we don't recommend tuning `l2_regularization`, as it does
+not significantly affect the generalization performance of gradient boosting
+models.
 
-The ``l2_regularization`` parameter is a regularizer on the loss function and
-corresponds to :math:`\lambda` in equation (2) of [XGBoost]_.
+|details-start|
+**Leaf weights in regression and classification**:
+|details-split|
+
+**Regression**: In the case of regression, the leaf weight is a continuous value
+that directly contributes to the model's prediction for a given input. The final
+prediction is the sum of the base prediction and the contributions from each
+tree. As regression gradient boosting models grow one tree per iteration, then
+:math:`k` goes from 1 to `max_iter`.
```

Comment:
The leaf weight does not really "directly" contribute to the prediction because the sum of the weights can be transformed by an inverse link function (depending on the choice of the loss).

> In the case of regression, the leaf weight is a continuous value
that contributes to the model's prediction for a given input that ends up in a given leaf. The final prediction is the sum of the base prediction and the contributions from each tree. The result of that sum can finally transformed by the inverse link function depending on the choice of the loss function: the exponential function is applied if the loss of the regressor is Poisson loss for instance.